### ISSUE TITLE
How the parameters of decay_rate & decay_steps are taken into consideration while computing InverseTimeDecay()

### ISSUE THREAD

**Issue by NikosSpanos** [tensorflow/tensorflow#46071]
During the last couple of days, I am experimenting with the different schedulers of learning rate decay offered by Keras (link [here][1]). Specifically, I have been using *InverseTimeDecay: A LearningRateSchedule that uses an inverse time decay schedule.* This type of scheduler uses the following two parameters:
1) decay_steps
2) decay_rate

*Note that those two parameters are also used by ```class ExponentialDecay```*

I would like to understand how those two parameters are initialized and taken into consideration by the InverseTimeDecay class. So far, what I know is that the following formula is used (based on the documentation):

```
def decayed_learning_rate(step):
  return initial_learning_rate / (1 + decay_rate * step / decay_step)
```

The parameter initialization I use:

```
def optimizer_adam_v2(hparams):
    # Inverse Time Decay
    initial_learning_rate = 0.001
    decay_steps = int(np.ceil(len(x_train)//32))*10 #len(x_train) = 31334, and 10 represents the epoch. So after 10 epochs or similarly 9791 steps the decay rate will be applied.
    decay_rate = 0.5
    learning_rate_fn = keras.optimizers.schedules.InverseTimeDecay(initial_learning_rate, decay_steps, decay_rate)

    return keras.optimizers.Adam(learning_rate=learning_rate_fn)
```

Based on the values above and the formula, the ```decay_learning_rate``` should be equal to

```
decay_learning_rate = 0.001 / (1 + 0.5 * step / 9791) = 0.0009999489353010264
# 9791 = int(np.ceil(31334//32))*10 #decay_steps parameter
```

When the training process begins the learning rate starts to drop by the decay_learning_rate at every step of the training. However, based on the [documentation][2] (	```decay_steps``` is "How often to apply decay"). the learning rate should start dropping after the training pass the value of the decay_steps. So after 9791 steps (approximately 10 epochs). Instead, the learning rate starts to drop from epoch 0. 

Why is this happening? Why the decay starts from epoch 0 and not from epoch 10 (9791 steps have passed).

P.S 1: Similar [question][3]

**[UPDATE] - 02.01.2021**

This is also observed with the Piecewise Constant Decay (link [here][3])

Piecewise constant decay is a similar scheduler to InverseTimeDecay with a much simpler application. The concept is that for every N number of steps the learning rate is constant and changes based on an array of values.

For example I have implemented a piecewise decay scheduler like below:

```
def optimizer_adam_v2():
    # PiecewiseConstantDecay
    step = tf.Variable(0, trainable=False)
    boundaries = [1000, 5000]
    values = [0.001, 0.0005, 0.00025]

    learning_rate_fn = keras.optimizers.schedules.PiecewiseConstantDecay(boundaries, values)

    # Later, whenever we perform an optimization step, we pass in the step.
    learning_rate_adam = learning_rate_fn(step)

    return keras.optimizers.Adam(learning_rate=learning_rate_adam)
```

In short, *based always on the relevant documentation written by Tensorflow experts*, what the code above does is to use a learning rate of 0.001 for the first 1000 steps, 0.0005 for the next 5000 steps (until 6000 steps completed), and the for the rest of the training use 0.00025 learning rate.
Since I monitored the learning rate of my scheduler I got the following result [here ](https://drive.google.com/file/d/1jvDzOZMB9YkuUxQ4I0yCapI5Hb9LaQM-/view?usp=sharing)in my Drive folder

With the indication ```lr``` I monitor the learning rate of the Adam optimizer. Until epoch 15 the learning rate didn't change whatsoever. Based on step calculation methodology, 15 epochs is approximately equal to ```len(x_train)//batch_size * epochs = (31335//32)*15 = 979*15 = 14,687 steps.```

After all, my question is still on the table. Why the learning rate of Adam does not obey the change rules of each scheduler.

You may find the implementation of the ```Piecewise constant decay``` scheduler on my colab notebook [here](https://colab.research.google.com/drive/1BsdrgZDtoDEASTikKa4LEVAz7IBzp3jw?usp=sharing)

  [1]: https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/schedules
  [2]: https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/schedules/InverseTimeDecay
  [3]: https://stackoverflow.com/questions/60029027/decay-parameter-of-adam-optimizer-in-keras

**Comment by ravikyram** [comment]
@NikosSpanos 

Please, fill [issue template](https://github.com/tensorflow/tensorflow/issues/new/choose).
Request you to share colab link or simple standalone code to reproduce the issue in our environment. It helps in localizing the issue faster. Thanks!

**Comment by NikosSpanos** [comment]
> @NikosSpanos
> 
> Please, fill [issue template](https://github.com/tensorflow/tensorflow/issues/new/choose).
> Request you to share colab link or simple standalone code to reproduce the issue in our environment. It helps in localizing the issue faster. Thanks!

I have added a google colab link to my question. Also, I have initially posted this question as a Documentation Issue (from this [list](https://github.com/tensorflow/tensorflow/issues/new/choose)) but I deleted any irrelevant content/text to my question.

**Comment by NikosSpanos** [comment]
> @NikosSpanos This is a stale issue. Is this still an issue for you?
> 
> As Keras moved to separate repo (i can't move the issue), can you please close here and open in keras repo https://github.com/keras-team/keras/issues
> 
> Keras team is focussed on the new repo mentioned above. Also, when you open new issue, please share a simple standalone code to reproduce the issue. Thanks!

I will close the issue here. I won't re-open it in Keras repo because I didn't  have the time to check any new updates on this matter. Probably it's resolved. But in any case if this issue come to my attention again I will re-open it on the new Keras repo. 

Thank you!
